src/main/java/sirius/db/jdbc/SmartQuery.java

```diff
@@ -320,8 +318,44 @@ protected Iterator<E> pullNextBlock() {
             return block.iterator();
         }
 
+        private SmartQuery<E> adjustQuery(SmartQuery<E> query) {
+            SmartQuery<E> adjusted = query.copy();
+
+            if (adjusted.limit > 0) {
+                throw new UnsupportedOperationException("SmartQuery doesn't allow 'limit' in streamBlockwise");
+            }
+            if (adjusted.skip > 0) {
+                throw new UnsupportedOperationException("SmartQuery doesn't allow 'skip' in streamBlockwise");
+            }
+
+            // we need to guarantee an absolute ordering
+            if (adjusted.distinct) {
+                // we have distinct fields, so we can easily create an absolute ordering, if it's not already there
+                if (!new HashSet<>(Tuple.firsts(adjusted.orderBys)).containsAll(adjusted.fields)) {
+                    Set<Mapping> missingOrderBys = new HashSet<>(adjusted.fields);
+                    Tuple.firsts(orderBys).forEach(missingOrderBys::remove);
+                    missingOrderBys.forEach(adjusted::orderAsc);
+                }
+            } else {
+                adjusted.orderAsc(BaseEntity.ID);
+            }
+
+            if (!adjusted.distinct && !fields.isEmpty()) {
+                // We SELECT a subset of the columns to optimize the network bandwidth.
+                // When pulling the next block, we need to continue exactly where we left of, so we need to SELECT
+                // at least all the fields from the ORDER BY clause.
+                Set<Mapping> allFields = new HashSet<>(adjusted.fields);
+                if (!allFields.containsAll(Tuple.firsts(adjusted.orderBys))) {
+                    allFields.addAll(Tuple.firsts(adjusted.orderBys));
+                }
+                adjusted.fields(allFields.toArray(Mapping[]::new));
```

Comment:
shoudln't this be a pre-allocated 0-sized array?

Reply:
don't we still need to add ID as ordering to guarantee stable sorting?

Reply:
The ID is added in line 340. Do you prefer this more explicitly?

Reply:
Since JDK11, there is [`Collection#toArray(IntFunction<T[]>)`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Collection.html#toArray(java.util.function.IntFunction)), which takes a reference to an array constructor instead. The only purpose of the generator function / 0-sized array / pre-sized array is, that the method can somehow construct an array of the correct type. Using the generator function seems to be the most ideomatic way to do this.